_posts/2018-07-31-tutorial.md

```diff
@@ -0,0 +1,1271 @@
+* `potential` is a `Comparable` type (typically a number)
+* `reduce` is a `(T, T) -> T`
+* `local` is the local value (T)
+* `null` is the default value, in case there is nothing to accumulate
+
+Example: Count the devices in the system, accumulating the information in source
+```java
+def countDevices(source) {
+  C(distanceTo(source), sum, 1, 0)
+}
+```
+
+#### T: evolve state
+```java
+T(initial, zero, decay)
+```
+* A timer from initial to zero with a decay rate.
+
+#### S: partition the network
+```java
+S(grain, metric)
+```
+
+* Elects a leader every grain, using the provided metric.
+* Returns a boolean (true if the node is a leader)
+* `grain` is a number with the partition size
+* `metric` is a `()->Field<Number>` that computes the distances
+* A timer from `initial` to `zero` with a `decay` rate.
+
+
+#### advanced/voronoi.pt
+```java
+module advanced:voronoi
+import protelis:coord:sparsechoice
+import protelis:coord:spreading
+let metric = () -> { self.nbrRange() };
+G(S(25, metric), 0, metric, v -> { v + metric.apply() })
+```
+
+* It’s one line of “true” code
+
+#### 14-voronoi.yml
+
+```yaml
+incarnation: protelis
+network-model:
+  type: EuclideanDistance
+  parameters: [10]
+displacements:
+  - in:
+      type: Circle
+      parameters: [500, 0, 0, 50]
+    programs:
+      -
+        - time-distribution: 1
+          program: advanced:voronoi
+        - program: send
+```
+
+* Run with:
+```bash
+./gradlew -Pfile=14-voronoi
+```
+* Try to move the nodes around, and see how the system adapts!
+
+### Limitations of the building block
+
+* Programming with building blocks is way easier than fiddling with `rep`s and `nbr`s by hand
+* ...but it is not yet where we’d like to stand!
+* Many “useful” behaviors can’t be captured
+  * as they are not self-stabilizing
+  * Yet, they can be of use in many situations
+    * Think of gossip, for instance
+* It’s very hard to get a grasp of the global situation
+  * If there is an information spread from multiple sources, only
+  * information from the closest would make it
+  * This is ok in many situations
+  * But not enough in others
+* Many algorithms recur very often
+  * They should stay in a library
+
+
+### Use `protelis-lang` and focus on the application code
+
+`protelis-lang` is a library<sup>1</sup> intended to:
+* Provide a collection of the most widely used distributed algorithms
+  * Contains over a hundred of them, from literature and other tools
+  * If you develop something new, contribution is very welcome :)
+* Shift the focus of the designer entirely on application code
+* Provide means to overcome the limitations of the building blocks
+  * Through a number of meta-algorithms that allow parallel aligned executions
+  * Mainly by means of the novel `multiInstance` algorithm
+`protelis-lang` is part of the Protelis distribution and ready to use
+
+<sup>1</sup>First introduced this year at a SASO Workshop: the 2nd edition of eCAS — did you
```

Comment:
This is copypasted by the slides and needs to be rephrased an linked to the original work:
https://ieeexplore.ieee.org/document/8064092/
and to the presentation:
https://www.slideshare.net/DanySK/2nd-ecas-workshop-on-engineering-collective-adaptive-systems